buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java

```diff
@@ -1373,12 +1377,17 @@ protected final void checkIndex(int index, int fieldLength) {
     }
 
     final void checkIndex0(int index, int fieldLength) {
-        if (isOutOfBounds(index, fieldLength, capacity())) {
-            throw new IndexOutOfBoundsException(String.format(
-                    "index: %d, length: %d (expected: range(0, %d))", index, fieldLength, capacity()));
+        int capacity = capacity();
```

Comment:
1) It moves the error message creation to a separate method [to help inlining](https://wiki.openjdk.java.net/display/HotSpot/Inlining)
2) It calls capacity() once instead of twice to reduce the number of virtual method calls
  - Since capacity() is virtual, the JVM can't realize it won't change